File: test/language/templates/java/templates/SubtypeTemplateWithBuiltinTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.subtype_template_with_builtin.SubtypeTemplateWithBuiltin;
import templates.subtype_template_with_builtin.TestStructure_uint32;

public class SubtypeTemplateWithBuiltinTest
{
    @Test
    public void readWrite() throws IOException
    {
        SubtypeTemplateWithBuiltin subtypeTemplateWithBuiltin = new SubtypeTemplateWithBuiltin();
        subtypeTemplateWithBuiltin.setTest(new TestStructure_uint32(13));

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        subtypeTemplateWithBuiltin.write(writer);

        final BitStreamReader reader = new ByteArrayBitStreamReader(
                writer.toByteArray(), writer.getBitPosition());
        SubtypeTemplateWithBuiltin readSubtypeTemplateWithBuiltin = new SubtypeTemplateWithBuiltin(reader);
        assertTrue(subtypeTemplateWithBuiltin.equals(readSubtypeTemplateWithBuiltin));
    }
}

File: test/language/templates/cpp98/SubtypeTemplateWithBuiltinTest.cpp
#include "gtest/gtest.h"

#include "templates/subtype_template_with_builtin/SubtypeTemplateWithBuiltin.h"

namespace templates
{
namespace subtype_template_with_builtin
{

TEST(SubtypeTemplateWithBuiltinTest, readWrite)
{
    SubtypeTemplateWithBuiltin subtypeTemplateWithBuiltin;
    subtypeTemplateWithBuiltin.getTest().setValue(13);

    zserio::BitStreamWriter writer;
    subtypeTemplateWithBuiltin.write(writer);
    size_t bufferSize = 0;
    const uint8_t* buffer = writer.getWriteBuffer(bufferSize);

    zserio::BitStreamReader reader(buffer, bufferSize);
    SubtypeTemplateWithBuiltin readSubtypeTemplateWithBuiltin(reader);

    ASSERT_TRUE(subtypeTemplateWithBuiltin == readSubtypeTemplateWithBuiltin);
}

} // namespace subtype_template_with_builtin
} // namespace templates
